crates/dslab-core/src/state.rs

```diff
@@ -74,9 +76,53 @@ impl SimulationState {
         }
     }
 
+    pub fn add_ordered_event<T>(&mut self, data: T, src: Id, dest: Id, delay: f64) -> EventId
+    where
+        T: EventData,
+    {
+        if !self.can_add_ordered_event(delay) {
+            panic!("Event order is broken! Ordered events should be added in non-decreasing order of their time.");
+        }
+        let last_time = self.ordered_events.back().map_or(f64::MIN, |x| x.time);
+        let event_id = self.event_count;
+        let event = Event {
+            id: event_id,
+            time: last_time.max(self.clock + delay),
+            src,
+            dest,
+            data: Box::new(data),
+        };
+        if delay >= 0. {
+            self.ordered_events.push_back(event);
+            self.event_count += 1;
+            event_id
+        } else {
+            log_incorrect_event(event, &format!("negative delay {}", delay));
+            panic!("Event delay is negative! It is not allowed to add events from the past.");
+        }
+    }
+
+    pub fn can_add_ordered_event(&self, delay: f64) -> bool {
+        if let Some(evt) = self.ordered_events.back() {
+            if delay + self.clock < evt.time - 1e-12 {
```

Comment:
Комментарий чтобы помнить

```suggestion
            // small epsilon is used to account for floating-point errors
            if self.clock + delay < evt.time - 1e-12 {
```

Reply:
Добавил